packages/celotool/src/e2e-tests/utils.ts

```diff
@@ -339,3 +284,111 @@ export function getContext(gethConfig: GethRunConfig, verbose: boolean = verbose
+  const numProxies = proxyInstances.length
+
+  const proxyPrivateKeys = getPrivateKeysFor(AccountType.PROXY, mnemonic, numProxies)
+  const proxyEnodes = proxyPrivateKeys.map((x: string, i: number) => {
+    const privateKey = proxyInstances[i].privateKey || x
+    return [
+      proxyInstances[i].name,
+      getEnodeAddress(privateKeyToPublicKey(privateKey), '127.0.0.1', proxyInstances[i].proxyport!),
+      getEnodeAddress(privateKeyToPublicKey(privateKey), '127.0.0.1', proxyInstances[i].port),
+    ]
+  })
+
+  let validatorIndex = 0
+  let proxyIndex = 0
+
+  for (const instance of gethConfig.instances) {
+    // Non proxied validators and proxies should connect to the bootnode
+    if (!instance.isProxied) {
+      if (gethConfig.useBootnode) {
+        instance.bootnodeEnode = bootnodeEnode
+      }
+    } else {
+      // Proxied validators should connect to only the proxy
+      // Find this proxied validator's proxy
+      const proxyEnode = proxyEnodes.filter((x: any) => x[0] === instance.proxy)
+
+      if (proxyEnode.length !== 1) {
+        throw new Error('proxied validator must have exactly one proxy')
+      }
+
+      instance.proxies = [proxyEnode[0][1]!, proxyEnode[0][2]!]
+    }
+
+    // Set the private key for the validator or proxy instance
+    if (instance.validating) {
+      instance.privateKey = instance.privateKey || validatorPrivateKeys[validatorIndex]
+      validatorIndex++
+    } else if (instance.isProxy) {
+      instance.privateKey = instance.privateKey || proxyPrivateKeys[proxyIndex]
+      proxyIndex++
+    }
+  }
+
+  // The proxies will need to know their proxied validator's address
+  for (const instance of gethConfig.instances) {
+    if (instance.isProxy) {
+      const proxiedValidator = gethConfig.instances.filter(
+        (x: GethInstanceConfig) => x.proxy === instance.name
+      )
+      if (proxiedValidator.length !== 1) {
+        throw new Error('proxied validator must have exactly one proxy')
+      }
+
+      instance.proxiedValidatorAddress = privateKeyToAddress(proxiedValidator[0].privateKey!)
+    }
+  }
+}
+
+export async function jsonRpc(web3: Web3, method: string, params: any[] = []): Promise<any> {
+  return new Promise((resolve, reject) => {
+    web3.currentProvider.send(
+      {
+        jsonrpc: '2.0',
+        method,
+        params,
+        // salt id generation, milliseconds might not be
+        // enough to generate unique ids
+        id: new Date().getTime() + Math.floor(Math.random() * (1 + 100 - 1)),
+      },
+      // @ts-ignore
+      (err: any, result: any) => {
+        if (err) {
+          return reject(err)
+        }
+        return resolve(result)
+      }
+    )
+  })
+}
+
+export async function restartInstance(
+  gethConfig: GethRunConfig,
+  instance: GethInstanceConfig,
+  removeNodeKey: boolean = false,
+  verbose: boolean = false
+) {
+  await killInstance(instance)
+  await restoreDatadir(gethConfig.runPath, instance)
```

Comment:
Possibly a race condition where the data directory is restored before the instance is dead. Something like:
```ts
while ((await spawnCmd('kill', ['-0', instance.pid.toString()]) === 0) {
  await sleep(0.25)
}
```
Could help.